File: zoebot_golang/internal/storage/redis.go
// Package storage provides Redis persistence for ZoeBot.
package storage

import (
	"context"
	"encoding/json"
	"fmt"
	"log"
	"sync"
	"time"

	"github.com/redis/go-redis/v9"
	"github.com/zoebot/internal/config"
)

// TrackedPlayer represents a player being tracked for match notifications.
type TrackedPlayer struct {
	PUUID       string `json:"puuid"`
	LastMatchID string `json:"last_match_id"`
	ChannelID   string `json:"channel_id"`
	Name        string `json:"name"`
}

// RedisClient wraps go-redis client.
type RedisClient struct {
	client  *redis.Client
	enabled bool
	ctx     context.Context
}

// NewRedisClient creates a new Redis client using go-redis.
func NewRedisClient(cfg *config.Config) *RedisClient {
	redisURL := cfg.RedisURL
	if redisURL == "" {
		log.Println("Redis not configured (REDIS_URL missing), using memory only")
		return &RedisClient{enabled: false, ctx: context.Background()}
	}

	opt, err := redis.ParseURL(redisURL)
	if err != nil {
		log.Printf("Failed to parse REDIS_URL: %v", err)
		return &RedisClient{enabled: false, ctx: context.Background()}
	}

	// Optimize for serverless
	opt.PoolSize = 5
	opt.MinIdleConns = 1
	opt.DialTimeout = 5 * time.Second
	opt.ReadTimeout = 3 * time.Second
	opt.WriteTimeout = 3 * time.Second

	client := redis.NewClient(opt)
	ctx := context.Background()

	// Test connection
	if err := client.Ping(ctx).Err(); err != nil {
		log.Printf("Redis connection failed: %v", err)
		return &RedisClient{enabled: false, ctx: ctx}
	}

	log.Println("Redis connected successfully")
	return &RedisClient{
		client:  client,
		enabled: true,
		ctx:     ctx,
	}
}

// Get retrieves a value from Redis.
func (r *RedisClient) Get(key string) (string, error) {
	if !r.enabled {
		return "", nil
	}
	val, err := r.client.Get(r.ctx, key).Result()
	if err == redis.Nil {
		return "", nil
	}
	return val, err
}

// Set stores a value in Redis (no expiration).
func (r *RedisClient) Set(key string, value string) error {
	if !r.enabled {
		return nil
	}
	return r.client.Set(r.ctx, key, value, 0).Err()
}

// Delete removes a key from Redis.
func (r *RedisClient) Delete(key string) error {
	if !r.enabled {
		return nil
	}
	return r.client.Del(r.ctx, key).Err()
}

// TrackedPlayersStore manages tracked players persistence.
type TrackedPlayersStore struct {
	redis   *RedisClient
	key     string
	players map[string]*TrackedPlayer
	mu      sync.RWMutex
}

// NewTrackedPlayersStore creates a new tracked players store.
func NewTrackedPlayersStore(redis *RedisClient, key string) *TrackedPlayersStore {
	return &TrackedPlayersStore{
		redis:   redis,
		key:     key,
		players: make(map[string]*TrackedPlayer),
	}
}

// Load loads tracked players from Redis.
func (s *TrackedPlayersStore) Load() error {
	s.mu.Lock()
	defer s.mu.Unlock()

	data, err := s.redis.Get(s.key)
	if err != nil {
		return err
	}

	if data == "" {
		s.players = make(map[string]*TrackedPlayer)
		return nil
	}

	var players map[string]*TrackedPlayer
	if err := json.Unmarshal([]byte(data), &players); err != nil {
		return err
	}

	s.players = players
	log.Printf("Loaded %d players", len(s.players))
	return nil
}

// Save saves tracked players to Redis.
func (s *TrackedPlayersStore) Save() error {
	s.mu.RLock()
	data, err := json.Marshal(s.players)
	s.mu.RUnlock()

	if err != nil {
		return fmt.Errorf("failed to marshal players: %w", err)
	}

	if err := s.redis.Set(s.key, string(data)); err != nil {
		return err
	}

	return nil
}

// Get returns a tracked player by PUUID.
func (s *TrackedPlayersStore) Get(puuid string) (*TrackedPlayer, bool) {
	s.mu.RLock()
	defer s.mu.RUnlock()
	p, ok := s.players[puuid]
	return p, ok
}

// Set adds or updates a tracked player.
func (s *TrackedPlayersStore) Set(puuid string, player *TrackedPlayer) {
	s.mu.Lock()
	s.players[puuid] = player
	s.mu.Unlock()
}

// Delete removes a tracked player.
func (s *TrackedPlayersStore) Delete(puuid string) {
	s.mu.Lock()
	delete(s.players, puuid)
	s.mu.Unlock()
}

// GetAll returns all tracked players.
func (s *TrackedPlayersStore) GetAll() map[string]*TrackedPlayer {
	s.mu.RLock()
	defer s.mu.RUnlock()

	result := make(map[string]*TrackedPlayer, len(s.players))
	for k, v := range s.players {
		result[k] = v
	}
	return result
}

// GetByChannel returns all tracked players for a specific channel.
func (s *TrackedPlayersStore) GetByChannel(channelID string) []*TrackedPlayer {
	s.mu.RLock()
	defer s.mu.RUnlock()

	var result []*TrackedPlayer
	for _, p := range s.players {
		if p.ChannelID == channelID {
			result = append(result, p)
		}
	}
	return result
}

// Count returns the number of tracked players.
func (s *TrackedPlayersStore) Count() int {
	s.mu.RLock()
	defer s.mu.RUnlock()
	return len(s.players)
}

// UpdateLastMatch updates the last match ID for a player.
func (s *TrackedPlayersStore) UpdateLastMatch(puuid, matchID string) {
	s.mu.Lock()
	if p, ok := s.players[puuid]; ok {
		p.LastMatchID = matchID
	}
	s.mu.Unlock()
}

File: zoebot_python/utils/redis.py
"""
Redis Client for ZoeBot
Handles persistence with Upstash Redis REST API.
"""

import json
import requests
import logging
from typing import Any

from config import (
    UPSTASH_REDIS_REST_URL,
    UPSTASH_REDIS_REST_TOKEN,
    REDIS_KEY_TRACKED_PLAYERS,
)

logger = logging.getLogger(__name__)


class RedisClient:
    """Client for Upstash Redis REST API."""

    def __init__(self):
        self.url = UPSTASH_REDIS_REST_URL
        self.token = UPSTASH_REDIS_REST_TOKEN
        self.enabled = bool(self.url and self.token)

        if not self.enabled:
            logger.warning("Upstash Redis not configured, using in-memory storage")

    def _request(self, command: list) -> dict | None:
        """Make a request to Upstash Redis REST API."""
        if not self.enabled:
            return None

        try:
            response = requests.post(
                self.url,
                headers={"Authorization": f"Bearer {self.token}"},
                json=command,
                timeout=10,
            )
            if response.status_code == 200:
                return response.json()
            else:
                logger.error(f"Redis error: {response.status_code} - {response.text}")
                return None
        except Exception as e:
            logger.error(f"Redis request failed: {e}")
            return None

    def get(self, key: str) -> Any | None:
        """Get value from Redis."""
        result = self._request(["GET", key])
        if result and result.get("result"):
            try:
                return json.loads(result["result"])
            except json.JSONDecodeError:
                logger.error(f"Failed to parse Redis data for key: {key}")
        return None

    def set(self, key: str, value: Any) -> bool:
        """Set value in Redis."""
        result = self._request(["SET", key, json.dumps(value)])
        return result and result.get("result") == "OK"

    def delete(self, key: str) -> bool:
        """Delete key from Redis."""
        result = self._request(["DEL", key])
        return result is not None


# Singleton instance
redis_client = RedisClient()


# ═══════════════════════════════════════════════════════════════════════════════
# TRACKED PLAYERS PERSISTENCE
# ═══════════════════════════════════════════════════════════════════════════════


def load_tracked_players() -> dict:
    """Load tracked players from Redis."""
    data = redis_client.get(REDIS_KEY_TRACKED_PLAYERS)
    if data:
        logger.info(f"📂 Loaded {len(data)} tracked players from Redis")
        return data
    return {}


def save_tracked_players(tracked_players: dict) -> bool:
    """Save tracked players to Redis."""
    success = redis_client.set(REDIS_KEY_TRACKED_PLAYERS, tracked_players)
    if success:
        logger.info(f"💾 Saved {len(tracked_players)} tracked players to Redis")
    else:
        logger.warning("Failed to save tracked players to Redis")
    return success
